File: src-tauri/src/services/pacing_engine.rs
/// 节奏控制引擎
/// 
/// 负责：时长监控、疲劳预警、收束策略
/// 开发计划：W11

/// 节奏状态
#[derive(Debug, Clone, serde::Serialize)]
pub struct PacingState {
    pub elapsed_minutes: i64,
    pub max_minutes: i32,
    pub remaining_minutes: i32,
    pub phase: PacingPhase,
}

/// 节奏阶段
#[derive(Debug, Clone, serde::Serialize)]
pub enum PacingPhase {
    /// 正常学习
    Normal,
    /// 预警（剩余 5 分钟）
    Warning,
    /// 收束（开始总结）
    WindingDown,
    /// 超时降级
    Overtime,
}

/// 评估当前节奏状态（骨架）
pub fn evaluate(elapsed_minutes: i64, max_minutes: i32) -> PacingState {
    let remaining = max_minutes as i64 - elapsed_minutes;
    let phase = if remaining <= 0 {
        PacingPhase::Overtime
    } else if remaining <= 2 {
        PacingPhase::WindingDown
    } else if remaining <= 5 {
        PacingPhase::Warning
    } else {
        PacingPhase::Normal
    };

    PacingState {
        elapsed_minutes,
        max_minutes,
        remaining_minutes: remaining.max(0) as i32,
        phase,
    }
}

File: server/app/services/pacing_engine.py
"""节奏控制引擎 - 从 src-tauri/src/services/pacing_engine.rs 移植

四段节奏：
- Normal      - 正常学习
- Warning     - 还剩 ≤ 5 分钟，开始暗示
- WindingDown - 还剩 ≤ 2 分钟，收束总结
- Overtime    - 超时，降级模式
"""
from dataclasses import dataclass
from enum import Enum


class PacingPhase(str, Enum):
    NORMAL = "normal"
    WARNING = "warning"
    WINDING_DOWN = "winding_down"
    OVERTIME = "overtime"


@dataclass
class PacingStatus:
    phase: PacingPhase
    elapsed_minutes: int
    remaining_minutes: int
    max_session_minutes: int
    message: str


def evaluate(elapsed_minutes: int, max_session_minutes: int) -> PacingStatus:
    """根据已用时长和上限判定当前节奏阶段"""
    remaining = max_session_minutes - elapsed_minutes

    if remaining <= 0:
        return PacingStatus(
            phase=PacingPhase.OVERTIME,
            elapsed_minutes=elapsed_minutes,
            remaining_minutes=0,
            max_session_minutes=max_session_minutes,
            message="时间到啦！今天先到这里吧，我们保持好节奏",
        )
    if remaining <= 2:
        return PacingStatus(
            phase=PacingPhase.WINDING_DOWN,
            elapsed_minutes=elapsed_minutes,
            remaining_minutes=remaining,
            max_session_minutes=max_session_minutes,
            message=f"还剩 {remaining} 分钟，我们来做个小总结",
        )
    if remaining <= 5:
        return PacingStatus(
            phase=PacingPhase.WARNING,
            elapsed_minutes=elapsed_minutes,
            remaining_minutes=remaining,
            max_session_minutes=max_session_minutes,
            message=f"还剩 {remaining} 分钟，抓紧最后一题吧",
        )

    return PacingStatus(
        phase=PacingPhase.NORMAL,
        elapsed_minutes=elapsed_minutes,
        remaining_minutes=remaining,
        max_session_minutes=max_session_minutes,
        message="保持节奏，继续加油",
    )
